src/resolvers/git.cr

```diff
@@ -41,23 +41,24 @@ module Shards
       end
     end
 
-    def specs(versions)
-      specs = {} of String => Spec
-
-      versions.each do |version|
-        refs = git_refs(version)
-        yaml = capture("git show #{refs}:#{SPEC_FILENAME}")
-        specs[version] = Spec.from_yaml(yaml).tap { |spec| spec.resolver = self }
-      rescue Error
+    def spec(version = nil)
+      spec = Spec.from_yaml(read_spec(version))
+      spec.resolver = self
+
+      if version && version =~ VERSION_REFERENCE
+        # In this case, we know that the version was looked up by git tag, so we
+        # validate the spec version against the tag version.
+        version = version.lstrip('v')
+        if spec.version != version
+          spec.mismatched_version = true
+        end
       end
 
-      specs
+      spec
     end
 
-    def spec?(version)
-      refs = git_refs(version)
-      yaml = capture("git show #{refs}:#{SPEC_FILENAME}")
-      Spec.from_yaml(yaml)
+    private def spec?(version)
+      spec(version)
     rescue Error
     end
```

Comment:
Shouldn't this land in `Resolver` instead?

Reply:
`spec?` is never used outside of the git resolver. That's why it's `private`, it's just a helper method.

Reply:
Please don't raise/rescue, it's needlessly expensive.

Even if only GitResolver uses `spec?` today, it doesn't mean that a new resolver won't benefit from it. It would also permit the good old idiom of `spec?` returning `nil` and `spec` raising if `spec?` returned `nil`. That's why `spec?` is the abstract method.

Reply:
This is the exact same implementation as before. It'd be better not to raise for sure but this isn't any worse, so I'd rather this is improved in another PR. 

The purpose of these refactors is to make sure the spec is parsed from yaml in only a single place in the code.

Reply:
And along with these refactors this method should land in its proper place at least, aside of raising or not.

Reply:
If it ends up being needed, it's easy to add it back. I'm not going to add an unused method to the other resolver for dubious unnamed future uses.